File: src/bin/0432-all-oone-data-structure.py
# @leetup=info id=432 lang=python3 slug=all-oone-data-structure



# @leetup=code


class Node:
    def __init__(self, count):
        self.count = count
        # For O(1), create only one node for all keys which have the same count.
        self.keys = set()
        self.next = None
        self.prev = None


class AllOne:
    def __init__(self):
        self.key_to_node = {}

        # double linked list
        # [left, min_node, ...., max_node, right]
        self.left = Node(0)
        self.right = Node(inf)
        self.left.next = self.right
        self.right.prev = self.left

    def inc(self, key: str) -> None:
        if key not in self.key_to_node:
            node = Node(0)
            node.keys = set()
            node.keys.add(key)
            self.insert_node(node, self.left)
            self.key_to_node[key] = node
        else:
            node = self.key_to_node[key]

        if node.count + 1 == node.next.count:
            # Add key to node.next
            node.next.keys.add(key)
            self.key_to_node[key] = node.next
        else:
            # insert new node
            new_node = Node(node.count + 1)
            new_node.keys.add(key)
            self.insert_node(new_node, node)
            self.key_to_node[key] = new_node

        # remove key from node
        node.keys.remove(key)
        if not node.keys:
            self.remove_node(node)

    def dec(self, key: str) -> None:
        assert key in self.key_to_node
        node = self.key_to_node[key]

        if node.count == 1:
            del self.key_to_node[key]
        elif node.count - 1 == node.prev.count:
            # Add key to node.prev
            node.prev.keys.add(key)
            self.key_to_node[key] = node.prev
        else:
            # insert new node
            new_node = Node(node.count - 1)
            new_node.keys.add(key)
            self.insert_node(new_node, node.prev)
            self.key_to_node[key] = new_node

        # remove key from node
        node.keys.remove(key)
        if not node.keys:
            self.remove_node(node)

    def insert_node(self, node, prev):
        # [ prev, nxt] => [prev, node, nxt]
        nxt = prev.next

        prev.next = node
        node.prev = prev

        node.next = nxt
        nxt.prev = node

    def remove_node(self, node):
        #  [prev, node, nxt] => [ prev, nxt]
        prev, next = node.prev, node.next
        prev.next, next.prev = next, prev

    def getMaxKey(self) -> str:
        max_node = self.right.prev
        if max_node.keys:
            res = max_node.keys.pop()
            max_node.keys.add(res)
            return res
        return ""

    def getMinKey(self) -> str:
        min_node = self.left.next
        if min_node.keys:
            res = min_node.keys.pop()
            min_node.keys.add(res)
            return res
        return ""


# @leetup=code

File: src/bin/0432-all-oone-data-structure.cpp
// @leetup=info id=432 lang=cpp slug=all-oone-data-structure

#include <algorithm>
#include <cassert>
#include <iostream>
#include <list>
#include <map>
#include <set>
#include <tuple>
#include <unordered_map>
#include <vector>

// @leetup=code

struct Node {
  int count;
  std::unordered_set<std::string> keys;
};

class AllOne {
 public:
  using list_pos = std::list<Node>::iterator;
  AllOne() {}

  void inc(std::string key) {
    auto it = map_.find(key);

    list_pos cur;

    if (it == map_.end()) {
      Node node{0, {key}};
      list_.push_front(node);
      cur = list_.begin();
    } else {
      cur = it->second;
    }

    list_pos next_pos = std::next(cur);

    if (next_pos != list_.end() && cur->count + 1 == next_pos->count) {
      // Add key to the next.
      next_pos->keys.insert(key);
      map_[key] = next_pos;
    } else {
      // Insert a new node.
      Node new_node{cur->count + 1, {key}};
      map_[key] = list_.insert(next_pos, new_node);
    }

    // Remove key from cur's keys
    cur->keys.erase(key);
    if (cur->keys.empty()) {
      list_.erase(cur);
    }
  }

  void dec(std::string key) {
    auto it = map_.find(key);
    assert(it != map_.end());

    list_pos cur = it->second;

    if (cur->count == 1) {
      map_.erase(key);
    } else {
      if (cur != list_.begin()) {
        list_pos prev_pos = std::prev(cur);
        if (cur->count - 1 == prev_pos->count) {
          // Add key to the prev.
          prev_pos->keys.insert(key);
          map_[key] = prev_pos;
        } else {
          // Insert a new node.
          Node new_node{cur->count - 1, {key}};
          map_[key] = list_.insert(cur, new_node);
        }
      } else {
        // Insert a new node at front.
        Node new_node{cur->count - 1, {key}};
        list_.push_front(new_node);
        map_[key] = list_.begin();
      }
    }

    // Remove key from cur's keys.
    cur->keys.erase(key);
    if (cur->keys.empty()) {
      list_.erase(cur);
    }
  }

  std::string getMaxKey() {
    if (list_.empty()) {
      return "";
    }
    return *(list_.back().keys.begin());
  }

  std::string getMinKey() {
    if (list_.empty()) {
      return "";
    }
    return *(list_.front().keys.begin());
  }

 private:
  std::list<Node> list_;
  std::unordered_map<std::string, list_pos> map_;
};

/**
 * Your AllOne object will be instantiated and called as such:
 * AllOne* obj = new AllOne();
 * obj->inc(key);
 * obj->dec(key);
 * string param_3 = obj->getMaxKey();
 * string param_4 = obj->getMinKey();
 */
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
